File: ruby/lib/algorithms/dfs.rb
# frozen_string_literal: true

require_relative '../data_structures/stack'

module DSA
  module Algorithms
    # Depth-First Search algorithm.
    # Time: O(V + E) where V = vertices, E = edges
    # Space: O(V)
    module DFS
      class << self
        # Perform iterative DFS traversal from a starting vertex.
        def traverse(graph, start)
          result = []
          visited = {}

          return result unless graph.vertices.include?(start)

          stack = DataStructures::Stack.new
          stack.push(start)

          until stack.empty?
            vertex = stack.pop

            next if visited[vertex]

            visited[vertex] = true
            result << vertex

            # Push neighbors in reverse order to visit them in natural order
            graph.neighbors(vertex).reverse_each do |neighbor|
              stack.push(neighbor) unless visited[neighbor]
            end
          end

          result
        end

        # Perform recursive DFS traversal from a starting vertex.
        def traverse_recursive(graph, start)
          result = []
          visited = {}

          return result unless graph.vertices.include?(start)

          dfs_recursive(graph, start, visited, result)
          result
        end

        # Find a path between two vertices using DFS.
        def find_path(graph, start, finish)
          vertices = graph.vertices
          return nil unless vertices.include?(start) && vertices.include?(finish)
          return [start] if start == finish

          visited = {}
          parent = { start => nil }
          stack = DataStructures::Stack.new
          stack.push(start)

          until stack.empty?
            vertex = stack.pop

            next if visited[vertex]

            visited[vertex] = true

            return reconstruct_path(parent, start, finish) if vertex == finish

            graph.neighbors(vertex).each do |neighbor|
              unless visited[neighbor]
                parent[neighbor] = vertex
                stack.push(neighbor)
              end
            end
          end

          nil
        end

        # Detect if the graph contains a cycle.
        def has_cycle?(graph)
          visited = {}
          recursion_stack = {}

          graph.vertices.each do |vertex|
            unless visited[vertex]
              return true if has_cycle_util?(graph, vertex, visited, recursion_stack)
            end
          end

          false
        end

        # Perform topological sort on a directed acyclic graph.
        def topological_sort(graph)
          return nil unless graph.directed

          visited = {}
          result = []
          recursion_stack = {}

          graph.vertices.each do |vertex|
            unless visited[vertex]
              return nil unless topological_sort_util(graph, vertex, visited, result, recursion_stack)
            end
          end

          result.reverse
        end

        private

        def dfs_recursive(graph, vertex, visited, result)
          visited[vertex] = true
          result << vertex

          graph.neighbors(vertex).each do |neighbor|
            dfs_recursive(graph, neighbor, visited, result) unless visited[neighbor]
          end
        end

        def has_cycle_util?(graph, vertex, visited, recursion_stack)
          visited[vertex] = true
          recursion_stack[vertex] = true

          graph.neighbors(vertex).each do |neighbor|
            if !visited[neighbor]
              return true if has_cycle_util?(graph, neighbor, visited, recursion_stack)
            elsif recursion_stack[neighbor]
              return true
            end
          end

          recursion_stack.delete(vertex)
          false
        end

        def topological_sort_util(graph, vertex, visited, result, recursion_stack)
          visited[vertex] = true
          recursion_stack[vertex] = true

          graph.neighbors(vertex).each do |neighbor|
            return false if recursion_stack[neighbor] # Cycle detected

            unless visited[neighbor]
              return false unless topological_sort_util(graph, neighbor, visited, result, recursion_stack)
            end
          end

          recursion_stack.delete(vertex)
          result << vertex
          true
        end

        def reconstruct_path(parent, start, finish)
          path = []
          current = finish

          while current
            path.unshift(current)
            current = parent[current]
          end

          path
        end
      end
    end
  end
end

File: scala/src/main/scala/algorithms/DFS.scala
package algorithms

import scala.collection.mutable

/**
 * Depth-first search algorithms.
 * Time: O(V + E)
 * Space: O(V)
 */
object DFS:
  def traverse[T](graph: Graph[T], start: T): List[T] =
    if !graph.hasVertex(start) then return List.empty

    val visited = mutable.Set.empty[T]
    val stack = mutable.Stack[T](start)
    val result = List.newBuilder[T]

    while stack.nonEmpty do
      val vertex = stack.pop()
      if !visited.contains(vertex) then
        visited += vertex
        result += vertex

        for neighbor <- graph.neighbors(vertex) do
          if !visited.contains(neighbor) then
            stack.push(neighbor)

    result.result()

  def traverseRecursive[T](graph: Graph[T], start: T): List[T] =
    if !graph.hasVertex(start) then return List.empty

    val visited = mutable.Set.empty[T]
    val result = List.newBuilder[T]

    def dfs(vertex: T): Unit =
      visited += vertex
      result += vertex
      for neighbor <- graph.neighbors(vertex) do
        if !visited.contains(neighbor) then
          dfs(neighbor)

    dfs(start)
    result.result()

  def findPath[T](graph: Graph[T], start: T, end: T): Option[List[T]] =
    if !graph.hasVertex(start) || !graph.hasVertex(end) then return None
    if start == end then return Some(List(start))

    val visited = mutable.Set.empty[T]
    val parent = mutable.Map.empty[T, T]
    val stack = mutable.Stack[T](start)

    while stack.nonEmpty do
      val vertex = stack.pop()
      if !visited.contains(vertex) then
        visited += vertex

        if vertex == end then
          return Some(reconstructPath(parent, start, end))

        for neighbor <- graph.neighbors(vertex) do
          if !visited.contains(neighbor) then
            parent(neighbor) = vertex
            stack.push(neighbor)

    None

  private def reconstructPath[T](parent: mutable.Map[T, T], start: T, end: T): List[T] =
    val path = List.newBuilder[T]
    var current = end
    while current != start do
      path += current
      current = parent(current)
    path += start
    path.result().reverse

  def hasCycle[T](graph: Graph[T]): Boolean =
    val visited = mutable.Set.empty[T]
    val recStack = mutable.Set.empty[T]

    def hasCycleUtil(vertex: T): Boolean =
      visited += vertex
      recStack += vertex

      for neighbor <- graph.neighbors(vertex) do
        if !visited.contains(neighbor) then
          if hasCycleUtil(neighbor) then return true
        else if recStack.contains(neighbor) then
          return true

      recStack -= vertex
      false

    for vertex <- graph.vertices do
      if !visited.contains(vertex) then
        if hasCycleUtil(vertex) then return true

    false

  def topologicalSort[T](graph: Graph[T]): Option[List[T]] =
    if !graph.directed then return None

    val visited = mutable.Set.empty[T]
    val recStack = mutable.Set.empty[T]
    val result = mutable.Stack.empty[T]

    def visit(vertex: T): Boolean =
      if recStack.contains(vertex) then return false
      if visited.contains(vertex) then return true

      visited += vertex
      recStack += vertex

      for neighbor <- graph.neighbors(vertex) do
        if !visit(neighbor) then return false

      recStack -= vertex
      result.push(vertex)
      true

    for vertex <- graph.vertices do
      if !visited.contains(vertex) then
        if !visit(vertex) then return None

    Some(result.toList)
